src/atem.ts

```diff
@@ -554,4 +605,91 @@ export class Atem extends EventEmitter {
 			delete this._sentQueue[trackingId]
 		}
 	}
+
+	/**
+	 * Helper method used by listVisibleInputs.
+	 * This got broken out into its own method because
+	 * it gets called multiple times, and gets called in a loop.
+	 * Breaking it out made listVisibleInputs much easier to read.
+	 */
+	private _calcActiveMeInputs (mode: 'program' | 'preview', meId: number): number[] {
+		const inputs = new Set<number>()
+		const meRef = this.state.video.getMe(meId)
+		inputs.add(meRef[mode === 'program' ? 'programInput' : 'previewInput'])
+
+		// Upstream Keyers
+		Object.values(meRef.upstreamKeyers).filter(usk => {
+			if (mode === 'program') {
+				return usk.onAir
+			}
+
+			// Pretty gross bitwise operations in this next line, be warned.
+			// TODO(Lange): This feels fragile but I'm unsure how to make it more robust.
+			return meRef.transitionProperties.selection & (2 ** (usk.upstreamKeyerId + 1))
+		}).forEach(usk => {
+			inputs.add(usk.fillSource)
+
+			// This is the only USK type that actually uses the cutSource.
+			if (usk.mixEffectKeyType === Enums.MixEffectKeyType.Luma) {
+				inputs.add(usk.cutSource)
+			}
+		})
+
+		// DSKs only show up on ME 1,
+		// so we only add them if that's the ME we are currently processing.
+		// TODO: Is this still true for the new Constellation ATEMs?
+		if (meId === 0) {
+			Object.values(this.state.video.downstreamKeyers).filter(dsk => {
+				if (mode === 'program') {
+					return dsk.onAir || dsk.inTransition
+				}
+
+				if (!dsk.properties) {
+					// Data isn't hydrated yet, we'll get 'em next time.
+					return
+				}
+
+				return dsk.properties.tie && !dsk.onAir
+			}).forEach(dsk => {
+				inputs.add(dsk.sources.fillSource)
+				inputs.add(dsk.sources.cutSource)
+			})
+		}
+
+		// Compute what sources are currently participating in a transition.
+		// We only care about this for PGM.
+		if (meRef.inTransition && mode === 'program') {
+			// The previewInput is always participating in the transition.
```

Comment:
Just realized this is not true and that I need to read `transitionProperties.selection` to determine when to set this.

Reply:
Should be fixed now.